datafusion/physical-expr/src/expressions/any.rs

```diff
@@ -0,0 +1,674 @@
+pub struct AnyExpr {
+    value: Arc<dyn PhysicalExpr>,
+    op: Operator,
+    list: Arc<dyn PhysicalExpr>,
+}
+
+impl AnyExpr {
+    /// Create a new Any expression
+    pub fn new(
+        value: Arc<dyn PhysicalExpr>,
+        op: Operator,
+        list: Arc<dyn PhysicalExpr>,
+    ) -> Self {
+        Self { value, op, list }
+    }
+
+    /// Compare for specific utf8 types
+    fn compare_utf8<T: OffsetSizeTrait>(
+        &self,
+        array: ArrayRef,
+        list: &ListArray,
+        negated: bool,
+        list_from_scalar: bool,
+    ) -> Result<ColumnarValue> {
+        let array = array
+            .as_any()
+            .downcast_ref::<GenericStringArray<T>>()
+            .unwrap();
+
+        if negated {
+            Ok(ColumnarValue::Array(Arc::new(neq_utf8(
+                array,
+                list,
+                list_from_scalar,
+            )?)))
+        } else {
+            Ok(ColumnarValue::Array(Arc::new(eq_utf8(
+                array,
+                list,
+                list_from_scalar,
+            )?)))
+        }
+    }
+
+    /// Get the left side of the binary expression
+    pub fn left(&self) -> &Arc<dyn PhysicalExpr> {
+        &self.value
+    }
+
+    /// Get the right side of the binary expression
+    pub fn right(&self) -> &Arc<dyn PhysicalExpr> {
+        &self.list
+    }
+
+    /// Get the operator for this binary expression
+    pub fn op(&self) -> &Operator {
+        &self.op
+    }
+
+    /// Compare for specific utf8 types
+    fn compare_bool(
+        &self,
+        array: ArrayRef,
+        list: &ListArray,
+        negated: bool,
+        list_from_scalar: bool,
+    ) -> Result<ColumnarValue> {
+        let array = array.as_any().downcast_ref::<BooleanArray>().unwrap();
+
+        if negated {
+            Ok(ColumnarValue::Array(Arc::new(neq_bool(
+                array,
+                list,
+                list_from_scalar,
+            )?)))
+        } else {
+            Ok(ColumnarValue::Array(Arc::new(eq_bool(
+                array,
+                list,
+                list_from_scalar,
+            )?)))
+        }
+    }
+}
+
+impl std::fmt::Display for AnyExpr {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{} {} ANY({})", self.value, self.op, self.list)
+    }
+}
+
+impl PhysicalExpr for AnyExpr {
+    /// Return a reference to Any that can be used for downcasting
+    fn as_any(&self) -> &dyn Any {
+        self
+    }
+
+    fn data_type(&self, _input_schema: &Schema) -> Result<DataType> {
+        Ok(DataType::Boolean)
+    }
+
+    fn nullable(&self, input_schema: &Schema) -> Result<bool> {
+        self.value.nullable(input_schema)
+    }
+
+    fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
+        let value = match self.value.evaluate(batch)? {
+            ColumnarValue::Array(array) => array,
+            ColumnarValue::Scalar(scalar) => scalar.to_array(),
+        };
```

Comment:
Thank you @ovr  -- this is looking quite cool.

I haven't reviewed this code super carefully but I wonder what you think about reusing more of the code for `InList`?

https://github.com/cube-js/arrow-datafusion/blob/binary-df-pr/datafusion/physical-expr/src/expressions/in_list.rs#L439

It seems like the implementations of `x IN (<expr>, <expr>)` and `x  ANY (<expr>)` are almost exactly the same except for the fact that the the `x = ANY(<expr>)` has the list as a single `<expr>` ?

The implementation of `get_set` needs to be different, but otherwise the rest of the code could probably be called directly.

I mention this as it is likely the `IN` / `NOT IN` are more optimized (use a hashset for testing, for example) as well as already written. 

Let me know what you think

Reply:
> I haven't reviewed this code super carefully but I wonder what you think about reusing more of the code for InList?

Probably, the nearest operator will be `ALL`, but it's not implemented right now. I tried to reuse the code of `InList`, but come to the decision that it will be easier/correct to combine it with ANY in the future.

> It seems like the implementations of x IN (<expr>, <expr>) and x ANY (<expr>) are almost exactly the same except for the fact that the the x = ANY(<expr>) has the list as a single <expr> ?

Nope, because it supports more operators: `<`, `<=`, `>`, `>=` (it's not implemented in this PR, but anyway).

> I mention this as it is likely the IN / NOT IN are more optimized (use a hashset for testing, for example) as well as already written.

Yes, but IN supports only the vector of scalars in DF. It's not the same as ANY, because it supports List (real column) or scalar.

For example `SELECT left, values FROM table`:

````
left | right LIst(Int64) - another logic is required
1.    | [1,2,3] - ArrayRef = PrimitiveArray<Int64> - another logic is requied
2.   |. [2,3,4]
3.   |. [4,5,6]
4.   |  [4,5,6]
5.   |. [4,5,6]
````

In the case of using the whole column, it is required to use another handling of interacting on values ☝️ . Probably here, will be more correct to go by steps:

- Support more operators
- Introduce `ALL` operator + extract the same logic with `ANY`.

Thanks